File: desktop-new/src/lib/ipc/terminal.ts
import { invoke } from "@tauri-apps/api/core"
import { listen, type UnlistenFn } from "@tauri-apps/api/event"

export async function terminalCreate(
  cols: number,
  rows: number,
): Promise<string> {
  return invoke<string>("terminal_create", { cols, rows })
}

export async function terminalCreateSsh(
  workspaceId: string,
  cols: number,
  rows: number,
): Promise<string> {
  return invoke<string>("terminal_create_ssh", { workspaceId, cols, rows })
}

export async function terminalWrite(
  sessionId: string,
  data: number[],
): Promise<void> {
  return invoke("terminal_write", { sessionId, data })
}

export async function terminalResize(
  sessionId: string,
  cols: number,
  rows: number,
): Promise<void> {
  return invoke("terminal_resize", { sessionId, cols, rows })
}

export async function terminalClose(sessionId: string): Promise<void> {
  return invoke("terminal_close", { sessionId })
}

export async function terminalListSessions(): Promise<string[]> {
  return invoke<string[]>("terminal_list")
}

interface TerminalOutputPayload {
  session_id: string
  data: number[]
}

interface TerminalExitPayload {
  session_id: string
}

export function onTerminalOutput(
  callback: (sessionId: string, data: Uint8Array) => void,
): Promise<UnlistenFn> {
  return listen<TerminalOutputPayload>("terminal:output", (event) => {
    callback(event.payload.session_id, new Uint8Array(event.payload.data))
  })
}

export function onTerminalExit(
  callback: (sessionId: string) => void,
): Promise<UnlistenFn> {
  return listen<TerminalExitPayload>("terminal:exit", (event) => {
    callback(event.payload.session_id)
  })
}

File: desktop-new/src-tauri/src/commands/terminal.rs
use crate::terminal::pty::PtyManager;
use std::sync::Arc;
use tauri::State;

#[tauri::command]
pub fn terminal_create(
    pty: State<'_, Arc<PtyManager>>,
    cols: u16,
    rows: u16,
) -> Result<String, String> {
    pty.create_session(cols, rows)
}

#[tauri::command]
pub fn terminal_create_ssh(
    pty: State<'_, Arc<PtyManager>>,
    workspace_id: String,
    cols: u16,
    rows: u16,
) -> Result<String, String> {
    pty.create_ssh_session(&workspace_id, cols, rows)
}

#[tauri::command]
pub fn terminal_write(
    pty: State<'_, Arc<PtyManager>>,
    session_id: String,
    data: Vec<u8>,
) -> Result<(), String> {
    pty.write_to_session(&session_id, &data)
}

#[tauri::command]
pub fn terminal_resize(
    pty: State<'_, Arc<PtyManager>>,
    session_id: String,
    cols: u16,
    rows: u16,
) -> Result<(), String> {
    pty.resize_session(&session_id, cols, rows)
}

#[tauri::command]
pub fn terminal_close(
    pty: State<'_, Arc<PtyManager>>,
    session_id: String,
) -> Result<(), String> {
    pty.close_session(&session_id)
}

#[tauri::command]
pub fn terminal_list(pty: State<'_, Arc<PtyManager>>) -> Result<Vec<String>, String> {
    pty.list_sessions()
}
